File: runtime/ruby/webhooks.rb
# frozen_string_literal: true

# Standard Webhooks verification (https://www.standardwebhooks.com).
# HMAC-SHA256 over `{id}.{timestamp}.{payload}` keyed by the base64-decoded
# secret (`whsec_` prefix optional); `webhook-signature` may carry several
# space-separated `v1,<base64>` candidates.

require "base64"
require "openssl"

module RedwoodModule
  class WebhookVerificationError < StandardError; end

  module Webhooks
    module_function

    # Raises WebhookVerificationError unless the payload is authentic.
    def verify(secret, payload, headers, tolerance_seconds: 300)
      raise WebhookVerificationError, "webhook secret must be a String" unless secret.is_a?(String)
      normalized = headers.to_h { |k, v| [k.to_s.downcase, v.is_a?(Array) ? v.first : v] }
      msg_id = normalized["webhook-id"]
      timestamp = normalized["webhook-timestamp"]
      signatures = normalized["webhook-signature"]
      if msg_id.to_s.empty? || timestamp.to_s.empty? || signatures.to_s.empty?
        raise WebhookVerificationError, "missing webhook-id, webhook-timestamp, or webhook-signature header"
      end

      # Tolerance must be finite and non-negative: an infinite tolerance
      # would silently disable replay protection.
      unless tolerance_seconds.is_a?(Numeric) && tolerance_seconds.finite? && tolerance_seconds >= 0
        raise WebhookVerificationError, "tolerance_seconds must be a finite non-negative number"
      end

      # The spec calls webhook-timestamp an integer Unix timestamp. Integer()
      # accepts "+1", whitespace, and underscores; require the wire grammar.
      unless /\A[0-9]+\z/.match?(timestamp.to_s)
        raise WebhookVerificationError, "webhook-timestamp is not an integer"
      end
      sent = Integer(timestamp)
      raise WebhookVerificationError, "webhook-timestamp is out of range" if sent > 2**63
      if (Time.now.to_f - sent).abs > tolerance_seconds
        raise WebhookVerificationError, "webhook-timestamp outside tolerance"
      end

      key = begin
        Base64.strict_decode64(secret.delete_prefix("whsec_"))
      rescue ArgumentError
        raise WebhookVerificationError, "webhook secret is not valid base64"
      end
      # Standard Webhooks symmetric keys are 24–64 bytes. A shorter key —
      # especially the zero-byte key from a bare "whsec_" — must never verify.
      unless key.bytesize.between?(24, 64)
        raise WebhookVerificationError, "decoded webhook secret must be 24-64 bytes, got #{key.bytesize}"
      end
      data = "#{msg_id}.#{timestamp}.#{payload}"
      expected = OpenSSL::HMAC.digest("SHA256", key, data)

      signatures.split(" ").each do |candidate|
        version, _, signature = candidate.partition(",")
        next unless version == "v1" && !signature.empty?

        provided = begin
          Base64.strict_decode64(signature)
        rescue ArgumentError
          next
        end
        return nil if provided.bytesize == expected.bytesize &&
                      OpenSSL.fixed_length_secure_compare(provided, expected)
      end
      raise WebhookVerificationError, "no matching v1 signature"
    end
  end
end

File: runtime/python/_webhooks.py
# Standard Webhooks verification (https://www.standardwebhooks.com).
# HMAC-SHA256 over `{id}.{timestamp}.{payload}` keyed by the base64-decoded
# secret (`whsec_` prefix optional); `webhook-signature` may carry several
# space-separated `v1,<base64>` candidates.
from __future__ import annotations

import base64
import hashlib
import hmac
import math
import re
import time
from typing import Mapping


class WebhookVerificationError(Exception):
    pass


def verify_webhook(
    secret: str,
    payload: bytes,
    headers: Mapping[str, str],
    *,
    tolerance_seconds: int = 300,
) -> None:
    """Raises WebhookVerificationError unless the payload is authentic."""
    normalized = {k.lower(): v for k, v in headers.items()}
    msg_id = normalized.get("webhook-id")
    timestamp = normalized.get("webhook-timestamp")
    signatures = normalized.get("webhook-signature")
    if not msg_id or not timestamp or not signatures:
        raise WebhookVerificationError(
            "missing webhook-id, webhook-timestamp, or webhook-signature header"
        )

    if not isinstance(secret, str):
        raise WebhookVerificationError("webhook secret must be a string")

    # Tolerance must be a finite non-negative number — and not a bool, which
    # would silently pass the int check.
    if (
        isinstance(tolerance_seconds, bool)
        or not isinstance(tolerance_seconds, (int, float))
        or not math.isfinite(tolerance_seconds)
        or tolerance_seconds < 0
    ):
        raise WebhookVerificationError("tolerance_seconds must be a finite non-negative number")

    # The spec calls webhook-timestamp an integer Unix timestamp. int()'s
    # broader grammar ("+1", " 1 ", "1_0") would disagree across SDKs, and a
    # float grammar would accept "nan" (comparisons always false — a replay-
    # window bypass). Bound it so float arithmetic can't overflow either.
    if not re.fullmatch(r"[0-9]+", str(timestamp)):
        raise WebhookVerificationError("webhook-timestamp is not an integer")
    sent = int(timestamp)
    if sent > 2**63:
        raise WebhookVerificationError("webhook-timestamp is out of range")
    if abs(time.time() - sent) > tolerance_seconds:
        raise WebhookVerificationError("webhook-timestamp outside tolerance")

    encoded = secret.removeprefix("whsec_")
    try:
        key = base64.b64decode(encoded, validate=True)
    except Exception:
        raise WebhookVerificationError("webhook secret is not valid base64") from None
    # Standard Webhooks symmetric keys are 24–64 bytes. A shorter key —
    # especially the zero-byte key from a bare "whsec_" — must never verify.
    if not 24 <= len(key) <= 64:
        raise WebhookVerificationError(f"decoded webhook secret must be 24-64 bytes, got {len(key)}")
    data = f"{msg_id}.{timestamp}.".encode() + payload
    expected = hmac.new(key, data, hashlib.sha256).digest()

    for candidate in signatures.split(" "):
        version, _, signature = candidate.partition(",")
        if version != "v1" or not signature:
            continue
        try:
            provided = base64.b64decode(signature)
        except Exception:
            continue
        if hmac.compare_digest(provided, expected):
            return
    raise WebhookVerificationError("no matching v1 signature")
